CovertInfiltration/Src/CovertInfiltration/Classes/X2TemplateModifier.uc

```diff
@@ -0,0 +1,126 @@
+	{
+		Schematic = X2SchematicTemplate(TemplateManager.FindItemTemplate(Item.SchematicName));
+		Template = TemplateManager.FindItemTemplate(Item.SchematicName);
+		if (Schematic != none)
+		{
+			// cycle through and modify each difficulty variant for this template
+			TemplateManager.FindDataTemplateAllDifficulties(Item.SchematicName, DifficultyVariants);
+			for (index = 0; index < DifficultyVariants.Length - 1; index++)
+			{
+				Schematic = X2SchematicTemplate(DifficultyVariants[index]);
+				
+				ValidIndex = index;
+				// if no cost value was set for this difficulty variant use next highest value
+				while (ValidIndex > 0 && Item.Costs[ValidIndex].ResourceCosts.Length == 0)
+				{
+					ValidIndex--;
+				}
+				
+				Schematic.bSquadUpgrade = false;
+				Schematic.OnBuiltFn = BuildItem;
+				Schematic.HideIfPurchased = '';
+				Schematic.bOneTimeBuild = false;
+				Schematic.Cost = Item.Costs[ValidIndex];
+
+				// only assign item template varaints on first pass or if nessecary
+				if (Template.bShouldCreateDifficultyVariants || index == 0)
+				{
+					Template = TemplateManager.FindItemTemplate(Schematic.ReferenceItemTemplate);
+
+					Template.HideInInventory = false;
+					Template.bInfiniteItem = false;
+				}
+			}
+		}
+		else if (Template != none && Template.CreatorTemplateName != '')
+		{
+			// TODO: create difficulty variants for these schematics
+			Schematic = X2SchematicTemplate(TemplateManager.FindItemTemplate(Template.CreatorTemplateName));
+			
+			// make sure we actually have a creator schematic for this template
+			if (Schematic != none)
+			{
+				Template.HideInInventory = false;
+				Template.bInfiniteItem = false;
+				Template.TradingPostValue = Item.TradingPostValue;
+				Template.CanBeBuilt = true;
+				Template.bInfiniteItem = false;
+				Template.BaseItem = 'None';
+				Template.UpgradeItem = 'None';
+				Template.CreatorTemplateName = 'None';
+				Template.Requirements = Schematic.Requirements;
+				Template.Cost = Item.Costs[0];
+			}
+		}
+	}
+}
+
+function BuildItem(XComGameState NewGameState, XComGameState_Item ItemState)
+{
+	local X2ItemTemplateManager TemplateManager;
+	local XComGameState_HeadquartersXCom XComHQ;
+	local X2SchematicTemplate Schematic;
+	local X2ItemTemplate ItemTemplate;
+	local XComGameState_Item NewItem;
+
+	foreach NewGameState.IterateByClassType(class'XComGameState_HeadquartersXCom', XComHQ)
+	{
+		break;
+	}
+	
+	if (XComHQ == none)
+	{
+		XComHQ = XComGameState_HeadquartersXCom(`XCOMHISTORY.GetSingleGameStateObjectForClass(class'XComGameState_HeadquartersXCom'));
+		XComHQ = XComGameState_HeadquartersXCom(NewGameState.CreateStateObject(class'XComGameState_HeadquartersXCom', XComHQ.ObjectID));
+		NewGameState.AddStateObject(XComHQ);
+	}
+
+	TemplateManager = class'X2ItemTemplateManager'.static.GetItemTemplateManager();
+	
+	Schematic = X2SchematicTemplate(ItemState.GetMyTemplate());
+	ItemTemplate = TemplateManager.FindItemTemplate(Schematic.ReferenceItemTemplate);
```

Comment:
Uhhh, this is really weird. Why is `ItemState.GetMyTemplate() != ItemTemplate`? Poor choice of variable names?

Reply:
it is .. in a round about kind of way it needs to pull the itemtemplate from the schematic template's ReferenceItemTemplate method or it doesnt work properly (dont know enough about templates to explain this functionality)